cpp/src/cider/common/base/StringRef.h

```diff
@@ -0,0 +1,279 @@
+ * For more information, see hash_map_string_2.cpp
+ */
+inline bool compareSSE2(const char* p1, const char* p2) {
+  return 0xFFFF == _mm_movemask_epi8(_mm_cmpeq_epi8(
+                       _mm_loadu_si128(reinterpret_cast<const __m128i*>(p1)),
+                       _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2))));
+}
+
+inline bool compareSSE2x4(const char* p1, const char* p2) {
+  return 0xFFFF ==
+         _mm_movemask_epi8(_mm_and_si128(
+             _mm_and_si128(
+                 _mm_cmpeq_epi8(_mm_loadu_si128(reinterpret_cast<const __m128i*>(p1)),
+                                _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2))),
+                 _mm_cmpeq_epi8(
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p1) + 1),
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2) + 1))),
+             _mm_and_si128(
+                 _mm_cmpeq_epi8(
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p1) + 2),
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2) + 2)),
+                 _mm_cmpeq_epi8(
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p1) + 3),
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2) + 3)))));
+}
+
+inline bool memequalSSE2Wide(const char* p1, const char* p2, size_t size) {
+  /** The order of branches and the trick with overlapping comparisons
+   * are the same as in memcpy implementation.
+   */
+  if (size <= 16) {
+    // TODO(Deegue): Add 17..24 bytes support
+    if (size >= 8) {
+      // Chunks of 8..16 bytes.
+      return unalignedLoad<uint64_t>(p1) == unalignedLoad<uint64_t>(p2) &&
+             unalignedLoad<uint64_t>(p1 + size - 8) ==
+                 unalignedLoad<uint64_t>(p2 + size - 8);
+    } else if (size >= 4) {
+      // Chunks of 4..7 bytes.
+      return unalignedLoad<uint32_t>(p1) == unalignedLoad<uint32_t>(p2) &&
+             unalignedLoad<uint32_t>(p1 + size - 4) ==
+                 unalignedLoad<uint32_t>(p2 + size - 4);
+    } else if (size >= 2) {
+      // Chunks of 2..3 bytes.
+      return unalignedLoad<uint16_t>(p1) == unalignedLoad<uint16_t>(p2) &&
+             unalignedLoad<uint16_t>(p1 + size - 2) ==
+                 unalignedLoad<uint16_t>(p2 + size - 2);
+    } else if (size >= 1) {
+      // A single byte.
+      return *p1 == *p2;
+    }
+    return true;
+  }
+
+  while (size >= 64) {
+    if (compareSSE2x4(p1, p2)) {
+      p1 += 64;
+      p2 += 64;
+      size -= 64;
+    } else {
+      return false;
+    }
+  }
+
+  switch (size / 16) {
+    case 3:
+      if (!compareSSE2(p1 + 32, p2 + 32))
+        return false;
+      [[fallthrough]];
+    case 2:
+      if (!compareSSE2(p1 + 16, p2 + 16))
+        return false;
+      [[fallthrough]];
+    case 1:
+      if (!compareSSE2(p1, p2))
+        return false;
+  }
+
+  return compareSSE2(p1 + size - 16, p2 + size - 16);

```

Comment:
`compareSSE2` will always compare 16 bytes? what will happen if call `memequalSSE2Wide(p1, p2, 20)`?

Reply:
Yes, `compareSSE2` will be called twice for a string of 20 bytes.

Reply:
oh, it will compare [0,15] and [4,19], right?

Reply:
yes